File: Leetcode/143.reorderList/143.cpp
 void reorderList(ListNode *head) 
 {
 	if (!head || !head -> next) return; 
 	//find the middle node o(n);
	ListNode * slow = head, fast = head;
	while (fast -> next && fast -> next -> next)
	{
		slow = slow -> next;
		fast = fast -> next -> next;	
	}
	//cur from the middle and reverse the second half;
	ListNode * head2 = slow -> next; 
	slow -> next = NULL;

	fast = head2 -> next;
	head2 -> next = NULL;
	while (fast)
	{
		slow = fast -> next;
		fast -> next = head2;
		head2 = fast;
		fast = slow;
	}
	slow = head;
	fast = head2;
//slow is the current tail of listnode, fast is next node  tmp the next next node; 
	while (slow)
	{
		ListNode* tmp = slow -> next;
		slow -> next = fast;
		slow = slow -> next;
		fast = tmp;
	}
}

File: Leetcode/143.reorderList/143.py
class Solution(object):
    def reorderList(self, head):
    	if not head or not head.next:
    		return
    	a, b = _spliList(head)
    	b = _reverseList(b)
    	head = _mergeLists(a, b)

    def _spliList(self, head):
    	fast = head
    	slow = head
    	while fast and fast.next:
    		slow = slow.next
    		fast = fast.next
    		fast = fast.next
    	middle = slow.next
    	slow.next = None
    	return head, middle

    def _reverseList(self,head):
    	last = None
    	currentNode = head
    	while currentNode:
    		nextNode = currentNode.next
    		currentNode.next = last
    		last = currentNode
    		currentNode = nextNode
    	return last

    def _mergeLists(self, a, b):
    	tail = a
    	head = a
    	a = a.next
    	while b :
    		tail.next = b
    		tail = tail.next
    		b = b.next
    		if a :
    			a, b = b, a
    	return head













